### ISSUE TITLE
Hybrid BGRU + Features Model

### ISSUE THREAD

**Issue by pravanjang** [pravanjang/stock-prediction-system#13]
**Code to Generate:**
Complete script `models/bgru_hybrid.py` implementing enhanced BGRU with features:

**Required Class:**
```python
class HybridBGRUModel:
    def __init__(self, sequence_length=60, n_price_features=5, n_static_features=50):
        """
        Hybrid Architecture:
        Path 1: BGRU processes sequential OHLCV data
        Path 2: Dense layers process static features (technical indicators, temporal)
        Fusion: Concatenate BGRU output + static features
        Output: Dense layers for classification
        """
        
    def build_model(self):
        """
        Architecture:
        
        [Sequential Path - OHLCV sequences]
        Input: [batch, 60, 5]
            ↓
        BGRU Layer 1: 128 units, bidirectional
            ↓
        Dropout: 0.3
            ↓
        BGRU Layer 2: 64 units, bidirectional
            ↓
        Dropout: 0.3
            ↓
        Output: [batch, 128] (concatenated forward + backward)
        
        [Static Path - Technical/Temporal/Price Action features]
        Input: [batch, 50+]
            ↓
        Dense: 64 units, ReLU
            ↓
        Dropout: 0.3
            ↓
        Dense: 32 units, ReLU
            ↓
        Output: [batch, 32]
        
        [Fusion Layer]
        Concat: [batch, 160] (128 from BGRU + 32 from static)
            ↓
        Dense: 64 units, ReLU
            ↓
        Dropout: 0.3
            ↓
        Dense: 32 units, ReLU
            ↓
        Dropout: 0.2
            ↓
        Output: 1 unit, Sigmoid
        """
    
    def prepare_data(self, df):
        """
        Prepares two types of inputs:
        1. Sequential OHLCV: shape [batch, 60, 5]
        2. Static features: shape [batch, n_features]
        """
    
    def train(self, train_df, val_df, epochs=50, batch_size=64):
        """
        Training with two-phase approach:
        Phase 1 (20 epochs): Train entire model with lr=0.001
        Phase 2 (30 epochs): Fine-tune with lr=0.0001
        """
    
    def save_model(self, path='models/checkpoints/bgru_hybrid.pt')
    def load_model(self, path)
    def predict(self, test_df)
```

**Training Configuration:**
- Optimizer: Adam
  - Phase 1: lr=0.001 (20 epochs)
  - Phase 2: lr=0.0001 (30 epochs) 
- Loss: Binary Cross-Entropy with class weights if imbalanced
- Batch size: 64
- Early stopping: patience=10
- Gradient clipping: max_norm=1.0

**Execution:**
```bash
python models/bgru_hybrid.py --train --data data/processed/train_final.csv --epochs 50
```

**Output:**
- `models/checkpoints/bgru_hybrid.pt` - trained model
- `models/checkpoints/hybrid_training_history.json` - metrics
- `models/logs/hybrid_training.log` - detailed logs

### 2.6 Feature Selection & Importance

**Code to Generate:**
Complete script `features/feature_selection.py`:

**Required Functions:**
```python
def calculate_permutation_importance(model, test_df, n_repeats=10)
def calculate_correlation_matrix(df)
def remove_multicollinear_features(df, vif_threshold=10)
def select_top_features(importance_dict, top_k=40)
def save_feature_importance_plot(importance_dict, save_path)
```

**Analysis to Perform:**
1. **Permutation Importance:**
   - Shuffle each feature, measure drop in accuracy
   - Rank features by importance
   - Save to `features/feature_importance.csv`

2. **Multicollinearity Check:**
   - Calculate VIF for all features
   - Remove features with VIF > 10
   - Save removed features to `features/removed_features.txt`

3. **Feature Selection:**
   - Keep top 40 most important features
   - Save selected feature list to `features/selected_features.txt`

**Visualization:**
- Plot top 20 features as horizontal bar chart
- Save to `features/feature_importance_plot.png`

**Execution:**
```bash
python features/feature_selection.py --model models/checkpoints/bgru_hybrid.pt --data data/processed/test_final.csv
```

**Output:**
- `features/feature_importance.csv`
- `features/selected_features.txt`
- `features/removed_features.txt`
- `features/feature_importance_plot.png`
